segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -1132,207 +1080,792 @@ public boolean supportsTruncation() {
+        private final List<ChunkNameOffsetPair> newReadIndexEntries = new ArrayList<>();
+        private int chunksAddedCount = 0;
+        private boolean isCommited = false;
+
+        private long traceId;
+        private Timer timer;
+
+        private String streamSegmentName;
+        private SegmentMetadata segmentMetadata;
+
+        private boolean isSystemSegment;
+
+        // Check if this is a first write after ownership changed.
+        private boolean isFirstWriteAfterFailover;
+
+        private ChunkMetadata lastChunkMetadata = null;
+        private ChunkHandle chunkHandle = null;
+        private int bytesRemaining;
+        private long currentOffset;
+
+        private boolean didSegmentLayoutChange = false;
+
+        WriteOperation(ChunkedSegmentStorage chunkedSegmentStorage, SegmentHandle handle, long offset, InputStream data, int length) {
+            this.handle = handle;
+            this.offset = offset;
+            this.data = data;
+            this.length = length;
+            this.chunkedSegmentStorage = chunkedSegmentStorage;
+        }
 
-                // Validate preconditions.
-                checkSegmentExists(streamSegmentName, segmentMetadata);
+        public CompletableFuture<Void> call() {
+            traceId = LoggerHelpers.traceEnter(log, "write", handle, offset, length);
+            timer = new Timer();
 
-                segmentMetadata.checkInvariants();
+            // Validate preconditions.
+            checkPreconditions();
+
+            streamSegmentName = handle.getSegmentName();
+            return tryWith(chunkedSegmentStorage.metadataStore.beginTransaction(handle.getSegmentName()),
+                    txn -> {
+                        didSegmentLayoutChange = false;
+
+                        // Retrieve metadata.
+                        return txn.get(streamSegmentName)
+                                .thenComposeAsync(storageMetadata -> {
+                                    segmentMetadata = (SegmentMetadata) storageMetadata;
+                                    // Validate preconditions.
+                                    checkState();
+
+                                    isSystemSegment = chunkedSegmentStorage.isStorageSystemSegment(segmentMetadata);
+
+                                    // Check if this is a first write after ownership changed.
+                                    isFirstWriteAfterFailover = segmentMetadata.isOwnershipChanged();
+
+                                    lastChunkMetadata = null;
+                                    chunkHandle = null;
+                                    bytesRemaining = length;
+                                    currentOffset = offset;
+
+                                    // Get the last chunk segmentMetadata for the segment.
+
+                                    return getLastChunk(txn)
+                                            .thenComposeAsync(v ->
```

Comment:
Is there another way to format this method? My head hurts :)

Reply:
mine too 👍